Day 2: “Using the Ancient Evils for Debugging” / Manuel Strehl (#207)
matuzo/HTMHell: hell/adventcalendar/2025/2/index.md

---
title: "Using the Ancient Evils for Debugging"
author: "Manuel Strehl"
author_bio: 'Manuel is a Germany-based web developer. Working in a small agency named <a href="https://kinetiqa.de">Kinetiqa</a> he is tasked with everything web that comes our way, from DB optimizations to accessibility testings. Manuel has been in this business long enough to show young developers his scars from the 2nd Browser War. In his little spare time he works on <a href="https://codepoints.net">codepoints.net</a>.'
date: 2025-12-02
author_links:
  - label: "Manuel’s Website"
    url: "https://manuel-strehl.de"
    link_label: "manuel-strehl.de"
  - label: "Manuel on Mastodon"
    url: "https://mastodon.social/@boldewyn"
    link_label: "@boldewyn@mastodon.social"
intro: "<p>There are unspeakable horrors in the depth of the HTML standard. We will take one of them today and unmystify it for our own use.</p>"
image: "advent25_2"
---

> Deep down in the dark voids of HTML specs long gone sleeps a terrifying thing. Imagine, if you will, a DOM node so mighty, that it can change the `content-type` of parts of the document. An HTML element that makes the parser tremble and withdraw, and that cannot be stopped even by its own end tag.

Reviewer: ```suggestion
Deep down in the dark voids of HTML specs long gone sleeps a terrifying thing. Imagine, if you will, a DOM node so mighty, that it can change the document’s `content-type` mid-stream. An HTML element that makes the parser tremble and withdraw, and that cannot be stopped even by its own end tag.
```

The wise people of W3C try to keep the knowledge of this terror away from the mere mortals’ eyes to spare us the danger of its madness. They advise us not to use the magic tag name that is the incantation for this ancient malice.

We will, of course, do exactly this today. We’ll take a deep look at the

```html
<plaintext>
```

element and what fun things we can use it for.

## A Quick Warning

This being said, I’d like to point out one important thing: Do not use this element in production. The HTML living standard is [quite clear about this](https://html.spec.whatwg.org/multipage/obsolete.html#plaintext):

> Elements in the following list are entirely obsolete, and must not be used by authors: [...] `plaintext`

So, what does `<plaintext>` do that earned it its place on HTML’s list of deprecated elements? In a nutshell, it ends the HTML parser and instructs the browser to interpret _everything_ following as plain text.

## What Do We Use this Power For?

> That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.

Reviewer: Just a thought, it felt to me like the `By the way...` thought ending the quick warning section felt better in terms of flow, and the usage section can then jump immediately into the use case of the article.

```suggestion
That is to be taken literally. Really everything, including any closing `</plaintext>` or `</html>`, will be printed as if a rogue, unclosed `<pre>` would suddenly go haywire and slurp up the rest of the page. By the way, this makes `<plaintext>` the only non-empty element that has no end tag at all.

## What Do We Use this Power For?
```

> On first sight that sounds like a really stupid superpower. On second sight, it still does. We look into how that element became part of HTML below. But now we will use it for one specific purpose: Debugging server-side code.

Reviewer: > On second sight, it still does

lol

> On first sight that sounds like a really stupid superpower. On second sight, it still does. We look into how that element became part of HTML below. But now we will use it for one specific purpose: Debugging server-side code.

Reviewer: > On second sight, it still does.

lol 😆

```suggestion
At first glance, that sounds like a really stupid superpower. At second glance, it still does. We look into how that element became part of HTML below. But first, let's put it to work for one specific purpose: debugging server-side code.
```

Author: I’ve kept the “sight”, but what I _actually_ wanted to use is “thought”. That sounds, however, even stranger.

> Of course, specialized debuggers like XDebug for PHP or built-in error pages like in Django take over the heavy lifting here. And even the good ol’ `print "<script>console.log('here!')</script>"` is often helpful. Those tools should be high up in your utility belt.

Reviewer: ```suggestion
Of course, specialized debuggers like XDebug for PHP or built-in error pages built-in error pages in frameworks like Django take over the heavy lifting here. And even the good ol’ `print "<script>console.log('here!')</script>"` is often helpful. Those tools should be high up in your utility belt.
```

> But imagine this: You are deep down in your code chasing an elusive bug that only affects some part of the HTML output, and you want to see at a quick glance on the rendered page, where this problem appears. The fastest way is to put a quick `<plaintext>` close to the offending place, reload the page, and presto! Just scan down to where the markup starts to show through.

Reviewer: Just reads a bit clunky and long, simplification suggestion:

```suggestion
But imagine this: you’re deep in your code, chasing an elusive bug that affects only part of the HTML output, and you want to spot on the rendered page exactly where it shows up.
```

> This is especially useful to access formatted debugging output. A `var_dump()` in PHP, for example. Or an `error.stack` stack trace in NodeJS. Slap a `<plaintext>` in front of it before writing it to the HTML output, so that the string is immediatelly readable:

Reviewer: ```suggestion
This is especially useful to access formatted debugging output. A `var_dump()` in PHP, for example. Or an `error.stack` stack trace in NodeJS. Slap a `<plaintext>` in front of it before sending it to the browser, making the string is immediately readable:
```

```php
<?php
# TODO delme!
echo '<plaintext>'; var_dump($strange_variable);
```

![A screenshot of the HTMHell website where the lower part shows a PHP variable output followed by the site’s markup instead of the rendered HTML](./debug_php.png)

> 

Reviewer: Great example, but a little brief – maybe a one-liner showing the same trick in JavaScript or Node would help generalize it?

## The History behind this Evil

> How ended this seemingly fringe feature up in all mainstream browsers? It was indeed there from the very beginning of HTML as this [historic W3C document](https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Tags.html) of 1992 proves:

Reviewer: Loving the folkloric literary touches 😆

> How ended this seemingly fringe feature up in all mainstream browsers? It was indeed there from the very beginning of HTML as this [historic W3C document](https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Tags.html) of 1992 proves:

Reviewer: Loving the folkloric literary touches 😆 

```suggestion
How ended this seemingly fringe feature up in all mainstream browsers? It was indeed there from the very beginning of HTML as this [historic W3C document](https://www.w3.org/History/19921103-hypertext/hypertext/WWW/MarkUp/Tags.html) from 1992 proves:
```

Just in case the backwards word order wasn't deliberate, `how did this ... end up in` would be the correct way around, but I think it works here and adds to the ancient horror style.

> **Plaintext**
>
> This tag indicates that all following text is to be taken litterally [!], up to the end of the file. Plain text is designed to be represented in the same way as example XMP text, with fixed width character and significant line breaks. Format:
>
> ```html
> <PLAINTEXT>
> ```
>
> This tag allows the rest of a file to be read efficiently without parsing. Its presence is an optimisation. There is no closing tag.

> This also tells us the reason for its invention. Back at the time Sir Tim Berners-Lee’s high-end NeXT PC that he used to write the first web browser had a quarter of the power of a hand-me-down 2009 smartphone. It was important to optimize wherever you could.

Reviewer: ```suggestion
This gives us a clue as to why it was introduced in the first place. Back at the time, Sir Tim Berners-Lee’s high-end NeXT PC that he used to write the first web browser had a quarter of the power of a hand-me-down 2009 smartphone. It was important to optimize wherever you could.
```

> Given that the early WWW was meant as a place to share scientific information, having a large blob of plain text as part of your fancy new HTML page was relatively common. The possibility to end the costly HTML parser and fall back to simply printing the remainder of the file as plain text was a powerful tool.

Reviewer: I know this is a hard ask, but it would be so cool to see an example screenshot here

> Such a feature isn’t unheard of in the world of computers, too. For example, the programming language Perl uses a [special marker](https://perldoc.perl.org/perldata#Special-Literals) to tell the Perl parser to stop processing the remainder of the file:

Reviewer: `world of computers` threw me a bit, I know what you mean here, but isn't HTML also `world of computers`? 

```suggestion
Such a feature isn’t unique to HTML, either. For example, the programming language Perl uses a [special marker](https://perldoc.perl.org/perldata#Special-Literals) to tell the Perl parser to stop processing the remainder of the file:
```

Author: Good point! Thank you!

```perl
print 'this is Perl code';
__END__
cout << 'this isn’t anymore';
```

(Perl programmers use this not for performance reasons but to embed additional data into their programs.)

Of course, nowadays, in the face of multi-megabyte JS payloads, this optimization has become completely unnecessary.

## How Safe Are We?

> But the element still _is_ available in all browsers. So we need to keep at least a passing knowledge of it at the back of our minds.

Reviewer: `keep ... passing knowledge` threw me a bit, might be a bit off idiomatically?

```suggestion
But the element _is_ still available in all browsers, so it's worth keeping a bit of a working knowledge of it at the back of our minds. 
```

Not sure...

Reviewer: Also, if you want to lean into the folklore thread, maybe `But the element still lurks in all browsers` 🧙

> To give you an example, how this feature could be mis-used, assume a comment function on a blog, where the commenter was able to smuggle in the string `<plaintext>`. As good developers we know never to trust users’ input, so we put the comment through a sanitizer. Let’s take a look at where things can go south from here.

Reviewer: Just a suggestion to sound more conversational/natural

```suggestion
To give you an example of how this feature could be misused, say a blog’s comment function lets someone smuggle in the string `<plaintext>`. As good developers we know never to trust users’ input, so we put the comment through a sanitizer. Let’s take a look at where things can go south from here.
```

We use the test string

```html
<p><b>hello<plaintext>world!</plaintext></b></p>
```

to check how several sanitizer libraries react to it.

### There Goes the Sanity!

> We run each sanitizer in its most minimal configuration that produces any output. This is by design: Sanitizers are security products. They should produce safe output by default.

Reviewer: ```suggestion
We run each sanitizer in the minimal configuration that still produces output. This is by design: sanitizers are security products. They should produce safe output by default.
```

> The results are in. And they are quite surprising. (Click on a name to read the horrifying details.)

Reviewer: The "horrifying details" setup made me expect that each sanitizer output would be individually alarming, but you don't really go into explaining what makes each of them bad. If I'm getting it right, further down it becomes clear that it's actually not that any of them are bad in any specific way, but the "horror" of it is that they're so inconsistent?

Maybe a clarification for that upfront would be good? As a reader, seeing "read these horrifying details", and then having the first example make me go "hmm, what was so bad about this?" made me feel a little stupid.

Maybe a small change along these lines would be good:

```suggestion
The results are in. And the inconsistencies are quite surprising. (Click on a name to read the horrifying details.)
```

<details>

> <summary>The new HTML Sanitizer API as implemented in Firefox</summary>

Reviewer: Just a formatting thought – maybe bold all of the `<summary>` sanitiser titles so there's a bit more hierarchy when they're open?

[developer.mozilla.org/en-US/docs/Web/API/Document/parseHTML_static](https://developer.mozilla.org/en-US/docs/Web/API/Document/parseHTML_static)

**Code:**

```js
console.log(Document.parseHTML(TEST_STRING).body.innerHTML);
```

> The approach of this API is to get the nesting correct again somehow according to the HTML5 parser spec. That is, close the `<p>` and `<b>` tags, then re-open the `<b>` tag as the spec suggests. The API does not deal with the special semantics of `<plaintext>` at all, though.

Reviewer: ```suggestion
The approach of this API is to somehow get the nesting conformant to the HTML5 parser spec. That is, close the `<p>` and `<b>` tags, then re-open the `<b>` tag as the spec suggests. The API does not deal with the special semantics of `<plaintext>` at all, though.
```

```js
Document.parseHTML(TEST_STRING, { sanitizer: { removeElements: []}}).body.innerHTML
```

</details>
<details>
<summary>HTML Tidy</summary>

[www.html-tidy.org/](https://www.html-tidy.org/)

**Code:**

```sh
echo -n "$TEST_STRING" | tidy
```

**Result:**

The venerable Tidy replaces the `<plaintext>` with a `<pre>`. This is creative.

```html
<p><b>hello</b></p>
<pre><b>world!</b></pre>
```

</details>
<details>
<summary>xss</summary>

[jsxss.com/](https://jsxss.com/)

**Code:**

```js
import xss from 'xss';
console.log(xss(TEST_STRING));
```

**Result:**

A well-known JavaScript-based sanitizer with special focus on XSS prevention escapes only the `<plaintext>` tags and leaves everything else in place.

```html
<p><b>hello&lt;plaintext&gt;world!&lt;/plaintext&gt;</b></p>
```

</details>
<details>
<summary>DOMPurify</summary>

[github.com/cure53/DOMPurify](https://github.com/cure53/DOMPurify)

**Code:**

```js
import { JSDOM } from 'jsdom';
import DOMPurify from 'dompurify';

const purify = DOMPurify(new JSDOM('').window);
console.log(purify.sanitize(TEST_STRING));
```

**Result:**

> The classic JS sanitizer chooses to remove the `<plaintext>` and all its “content”. (I put content in quotes, because _technically_ everything after the start tag would’ve been the `<plaintext>`’s content.) DOMPurify sees to it, that the elements are properly closed. The result is identical to the one of the Sanitizer API.

Reviewer: ```suggestion
The classic JS sanitizer chooses to remove the `<plaintext>` and all its “content”. (I put content in quotes, because _technically_ everything after the start tag would’ve been the `<plaintext>`’s content.) DOMPurify sees to it that the elements are properly closed. The result is identical to that of the Sanitizer API.
```

```html
<p><b>hello</b></p>
```

</details>
<details>
<summary>HTML Purifier</summary>

[htmlpurifier.org/](http://htmlpurifier.org/)

**Code:**

```php
<?php
$config = HTMLPurifier_Config::createDefault();
$purifier = new HTMLPurifier($config);
printf($purifier->purify(TEST_STRING));
```

**Result:**

The top dog in the PHP world takes a slightly different approach. It removes only the element itself. (Note the “world!” remaining intact.)

```html
<p><b>helloworld!</b></p>
```

</details>
<details>
<summary>Symfony HtmlSanitizer</summary>

[symfony.com/html-sanitizer](https://symfony.com/html-sanitizer)

**Code:**

```php
<?php
use Symfony\Component\HtmlSanitizer\HtmlSanitizer;
use Symfony\Component\HtmlSanitizer\HtmlSanitizerConfig;

$config = (new HtmlSanitizerConfig())->allowSafeElements();
$sanitizer = new HTMLSanitizer($config);
printf($sanitizer->sanitize(TEST_STRING));
```

**Result:**

> Symfony’s sanitizer has a fascinating way of moving tags around. Interesting, but at least we’ve got all elements properly closed, including the un-closeable `plaintext`.

Reviewer: ```suggestion
Symfony’s sanitizer has a fascinating way of moving tags around. Interesting – but at least we’ve got all elements properly closed, including the unclosable `plaintext`.
```

Author: I’ve kept the comma here. The “Interesting” isn’t meant to be emphasized more than a ”hm, ok” would be, because the actual punch line of the sentence is the `</plaintext>` closing tag.

```

**Result:**

Python developers who reach for this library will have everything but the `<b>` escaped.

```html
&lt;p&gt;<b>hello&lt;plaintext&gt;world!&lt;/plaintext&gt;</b>&lt;/p&gt;
```

</details>
<details>
<summary>OWASP Java HTML Sanitizer</summary>

[github.com/OWASP/java-html-sanitizer/](https://github.com/OWASP/java-html-sanitizer/)

**Code:**

```java
import org.owasp.html.PolicyFactory;
import org.owasp.html.Sanitizers;

public class Sanitize {
    public static void main(String[] args) {
        PolicyFactory policy = Sanitizers.FORMATTING.and(Sanitizers.LINKS);
        String safe = policy.sanitize(TEST_STRING);
        System.out.println(safe);
    }
}
```

**Result:**

The staple HTML sanitizer in the Java world escapes everything and does strange things to the end tags, but at least the `<plaintext>` is gone.

```html
<b>helloworld!&lt;/plaintext&gt;&lt;/b&gt;&lt;/p&gt;</b>
```

</details>
<details>
<summary>Ammonia as configured by nh3</summary>

[github.com/rust-ammonia/ammonia](https://github.com/rust-ammonia/ammonia), [nh3.readthedocs.io/](https://nh3.readthedocs.io/)

**Code:**

```py
import nh3
print(nh3.clean(TEST_STRING))
```

> This Rust-based sanitizer advertises its speed and conformance with the HTML spec.
> 
> **Result:**

Reviewer: ```suggestion
**Result:**

This Rust-based sanitizer advertises its speed and conformance with the HTML spec.
```

> However, the result is close but still different to what browsers will do. In this case, it’s the `<b>` tag that would not extend over the content of the `<plaintext>` element.

Reviewer: `still` already conveys the contrasting thought.

```suggestion
The result is close but still different to what browsers will do. In this case, it’s the `<b>` tag that would not extend over the content of the `<plaintext>` element.
```

```html
<p><b>hello</b></p><b>world!&lt;/plaintext&gt;&lt;/b&gt;&lt;/p&gt;</b>
```

</details>

---

> With **11** sanitizing methods we managed to produce **10** different outputs.

Reviewer: Just stylistic choice, take it or leave it!

```suggestion
With **11** sanitizing methods we managed to produce **10** different outputs!
```

Just think this is a really nice overview and deservers a little pause!

Just to be crystal clear here: we are not criticizing the result of any of these libraries. Each one has a good reason to do what they do. And each one serves a slightly different purpose.

> It emphasizes the point, though, that one should be absolutely sure about the aim of a chosen sanitizer and extent as to which it will change its input. Is the library for removing only potentially dangerous things but keeping as much HTML intact as possible? Is it to scrape all HTML off the string, or only escaping any HTML-special characters? The results will differ tremendously.

Reviewer: ```suggestion
But it does emphasize the point that one should be absolutely sure about the aim of a chosen sanitizer and the extent to which it will change its input. Is the library for removing only potentially dangerous things but keeping as much HTML intact as possible? Is it to scrape all HTML off the string, or only escape special characters? The results will differ tremendously.
```

We enter the danger zone when mixing several tools together without taking a cautious look first. For example, look at how DOMPurify and HTML Purifier would interact in a potentially hazardous way. DOMPurify would remove any `<plaintext>` including its content. A later check for any malicious payload would be negative.

> HTML Purifier on the other hand just strips the `<plaintext>` tag, while its content remains on the page. If we’d trust the previous DOMPurify result, we’d be surprised by sudden new content being placed verbatim in the HTML code.

Reviewer: Slightly better paragraph break placement for flow, maybe?

```suggestion
We enter the danger zone when mixing several tools together without taking a cautious look first.

For example, look at how DOMPurify and HTML Purifier would interact in a potentially hazardous way. DOMPurify would remove any `<plaintext>` including its content. A later check for any malicious payload would be negative. HTML Purifier on the other hand just strips the `<plaintext>` tag, while its content remains on the page. If we’d trust the previous DOMPurify result, we’d be surprised by sudden new content being placed verbatim in the HTML code.
```

> If one library is used for input validation and another one for output quoting, this is a [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) desaster waiting to happen, unless we know _exactly_ what we’re doing.

Reviewer: ```suggestion
If one library is used for input validation and another one for output quoting, that's a [cross-site scripting](https://en.wikipedia.org/wiki/Cross-site_scripting) disaster waiting to happen, unless we know _exactly_ what we’re doing.
```

Author: Oh, interesting! It’s “Desaster” in German, but “disaster” in English. Hadn’t noticed that consciously before.

## Letting the Evil Sleep Again

> In the case of `<plaintext>` itself we are most likely in a safe place, though. Since `<plaintext>` has built-in HTML escaping, doing something dangerous with it is severely limited. It takes quite rare a constellation of errors and oversights to appear together in order to run malicious code.

Reviewer: ```suggestion
In the case of `<plaintext>` itself we are most likely in a safe place. Since `<plaintext>` has built-in HTML escaping, doing something dangerous with it is severely limited. It takes quite rare a constellation of errors and oversights to appear together in order to run malicious code.
```

> Well, for the sake of the argument, let’s create such a constellation. Assume that you embed a Content-Security Policy on your site [in a `<meta>` element](https://w3c.github.io/webappsec-csp/#meta-element) instead of an HTTP header:

Reviewer: ```suggestion
Just for the sake of the argument, let’s create such a constellation. Assume that you embed a Content-Security Policy on your site [in a `<meta>` element](https://w3c.github.io/webappsec-csp/#meta-element) instead of an HTTP header:
```

```html
<meta http-equiv="Content-Security-Policy" content="script-src 'self'">
```

> This prevents loading 3rd party scripts sufficiently. If an attacker finds a possibility to load HTML prior to this element, they can nullify the CSP:

Reviewer: ```suggestion
This sufficiently prevents loading 3rd party scripts. If an attacker finds a way to load HTML prior to this element, they can nullify the CSP:
```

```html
<script src="https://example.com/malicious.js"></script>
<plaintext>
<meta http-equiv="Content-Security-Policy" content="script-src 'self'">
```

But again, for this to really have any effect, several things must come together:

- The attacker must be able to place HTML in the `<head>` (because CSP meta tags can only be used there)
- The CSP is not set via HTTP
- The complete remainder of the page is converted to `text/plain`, which makes this definitively not a stealthy attack

> So we can conclude: It is important to know about `<plaintext>`. But if we follow tried and tested security rules (for example the [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/)), we will remain safe from this ancient evil.

Reviewer: ```suggestion
So, to sum up – it's important to know about `<plaintext>`. But if we follow tried and tested security rules (for example the [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/)), we will remain safe from this ancient evil.
```

Reviewer: I do love that you mention "ancient evil" again to tie it all together. Maybe it'd be nice to go a bit further and add a nod to the fact that it is "ancient"?

I'd also add some detail to summarise the big point from the article – that sanitizers don't agree on how to handle `<plaintext>`.

```suggestion
To sum up – every era of the web leaves behind strange relics like `<plaintext>`, and it's important to know about the ghosts that haunt the platform. This one is especially mischievous: sanitizers don't agree on how to treat it, so it pays to know its quirks. Follow sound security practices (like those in [OWASP ASVS](https://owasp.org/www-project-application-security-verification-standard/)), and we'll remain safe from this ancient evil.
```

Might be going a tad too far into the metaphor, up to you 😜
